tensorflow_model_optimization/python/core/sparsity_tf2/pruner.py

```diff
@@ -211,4 +211,64 @@ def prune(self, optimizer, var, grad):
     mask = optimizer.get_slot(var, 'mask')
     threshold = optimizer.get_slot(var, 'threshold')
     self.update_masks([(var, mask, threshold)], step=optimizer.iterations)
-    self._mask_weight(var, mask)
+    self._apply_mask(var, mask)
+
+class LTHPruner(Pruner):
+  """
+  Implementation of Lottery Ticket Hypothesis experiments.
+  """
+
+  def __init__(self,
+      pruning_schedule=pruning_sched.ConstantSparsity(0.5, 0),
+      save_schedule=None,
+      block_size=(1,1),
+      block_pooling_type='AVG',
+  ):
+    """The logic for magnitude-based pruning weight tensors.
+
+    Args:
+      pruning_schedule: A `PruningSchedule` object that controls pruning rate
+        throughout training.
+      reload_schedule: A `PruningSchedule` object that controls reloading of weights
+      throughout training. Default same as pruning schedule.
+      save_schedule: A integer representing the weights for relloading after checkpointing in LTH experiments.
+      block_size: The dimensions (height, weight) for the block sparse pattern
+        in rank-2 weight tensors.
+      block_pooling_type: (optional) The function to use to pool weights in the
+        block. Must be 'AVG' or 'MAX'.
+    """
+    super().__init__(pruning_schedule, block_size, block_pooling_type)
+    self.reload_schedule = pruning_schedule
+    self.save_step = save_schedule if save_schedule else 0
+    if issubclass(self.reload_schedule, pruning_sched.PruningSchedule) \
+      and tf.math.less(self.reload_schedule.get_config()['config']['begin_step'], self.save_step):
+      raise ValueError("Reloading should not occur before initializations are saved.")
+
+  
+  def create_slots(self, optimizer, var):
+    base_dtype = var.dtype
+    optimizer.add_slot(var, 'mask', initializer='ones')
+    optimizer.add_slot(var, 'threshold', initializer=tf.zeros(shape=(), dtype=base_dtype))
+    optimizer.add_slot(var, 'original_initialization', initializer=var.read_value())
+
+  def _maybe_save_weights(self, optimizer, var):
+    """
+    Save weights right before the save iteration.
+    """
+    if tf.math.equal(self.save_step - 1, optimizer.iterations):
```

Comment:
Don't worry about it for this PR,

but we could consider splitting `prune` into a `before_weight_update`/`before_gradient_update`/`preprocess_gradient` method & a `maybe_update_masks`/`at_weight_update_end`/`at_gradient_update_end` method. (We may need to do a split like this anyway to sparsify the gradients)

Then it could save the initializations at the start of save_step, and reload them after the weight updates. (w/o dealing w/ off-by-ones or having to initialize to the read_value)